p2panda-rs/src/schema/schema.rs

```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::BTreeMap;
+use std::fmt;
+
+use crate::hash::Hash;
+use crate::message::{Message, MessageFields, MessageValue};
+use crate::schema::SchemaError;
+
+use cddl::lexer::Lexer;
+use cddl::parser::Parser;
+#[cfg(not(target_arch = "wasm32"))]
+use cddl::validate_cbor_from_slice;
+#[cfg(not(target_arch = "wasm32"))]
+use cddl::validator::cbor;
+
+/// CDDL types
+#[derive(Clone, Debug, Copy)]
+#[allow(missing_docs)]
+pub enum Type {
+    Bool,
+    Int,
+    Float,
+    Tstr,
+    Relation,
+}
+
+/// CDDL schema type string formats
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let cddl_type = match self {
+            Type::Bool => "bool",
+            Type::Int => "int",
+            Type::Float => "float",
+            Type::Tstr => "tstr",
+            Type::Relation => "tstr .regexp \"[0-9a-fa-f]{132}\"",
+        };
+        write!(f, "{}", cddl_type)
+    }
+}
+
+/// CDDL field types
+#[derive(Clone, Debug)]
+pub enum Field {
+    String(String),
```

Comment:
Some of these are neither used nor tested as of now. Is it good to have them here in case we want to use them or should we restrict `Field` to only cover the field types we actually use?

Reply:
Hmmm, yeh, you're right. Were you thinking to remove any we don't currently use? I think that's a good move, it's very easy to implement the ones we need as we use them. I'll do that.